Saule/Serialization/ResourceSerializer.cs

```diff
@@ -66,71 +69,80 @@ public JObject Serialize(JsonSerializer serializer)
+            var dataArray = token as IEnumerable<object>;
 
             // single thing, just serialize it
             if (dataArray == null)
             {
-                return token is JObject ? serializeObj((JObject)token) : null;
+                return serializeObj(token);
             }
 
             // serialize each element separately
             var data = new JArray();
-            foreach (var obj in dataArray.OfType<JObject>())
+            foreach (var obj in dataArray)
             {
                 data.Add(serializeObj(obj));
             }
 
             return data;
         }
 
-        private static JToken SerializeMinimalData(IDictionary<string, JToken> properties, ApiResource resource)
+        private static JToken SerializeMinimalData(object properties, ApiResource resource)
         {
             var data = new JObject
             {
                 ["type"] = resource.ResourceType.ToDashed(),
-                ["id"] = EnsureHasId(properties, resource)
+                ["id"] = JToken.FromObject(EnsureHasId(properties, resource))
             };
 
             return data;
         }
 
-        private static JToken SerializeAttributes(IDictionary<string, JToken> properties, ApiResource resource)
+        private static JToken SerializeAttributes(object properties, ApiResource resource)
         {
             var attributes = new JObject();
             foreach (var attr in resource.Attributes)
             {
                 var value = GetValue(attr.Name, properties);
                 if (value != null)
                 {
-                    attributes.Add(attr.Name, value);
+                    attributes.Add(attr.Name, JToken.FromObject(value));
                 }
             }
 
             return attributes;
         }
 
-        private static JToken GetValue(string name, IDictionary<string, JToken> properties)
+        private static object GetValue(string name, dynamic obj)
         {
-            return properties[name.ToPascalCase()];
+            try
+            {
+                var propertyName = name.ToPascalCase();
+                return obj.Send(propertyName);
+            }
+            catch
+            {
+                return null;
+            }
+
         }
 
-        private static JToken GetId(IDictionary<string, JToken> properties, ApiResource resource)
+        private static object GetId(object properties, ApiResource resource)
         {
             return GetValue(resource.IdProperty, properties);
         }
 
-        private static JToken EnsureHasId(IDictionary<string, JToken> properties, ApiResource resource)
+        private static string EnsureHasId(object properties, ApiResource resource)
         {
             var id = GetId(properties, resource);
             if (id == null)
             {
                 throw new JsonApiException(ErrorType.Server, "Resources must have an id");
             }
 
-            return id;
+            return id.ToString();
```

Comment:
It's better to do `.ToString()` here than to cast to string, cause of weird primary key types (like guids --  /cc @wozbo)